### PR TITLE
Persistent queue

### PR DESCRIPTION
**Pull request by chme**
This pr changes the queue/current playlist handling to use a db table instead of an in memory queue. Combined with the (existing) config option "clear_queue_on_stop_disable" in the mpd section it allows to keep the queue/current playlist between restarts of forked-daapd (requested by @LordMyschkin).

It also decouples the queue from the files db table which should make it possible to update metadata for a stream (internet radio or pipe) without the need to modify the library/files table (updating the queue table should be sufficient - i will look into this). 
It is also a preliminary step to make it possible to add items to the queue that are not in the library. I hope in combination with the integration of the spotify web api, it enables searching for artist/album/songs in the spotify catalogue and adding them to the queue without the need to first add them to a playlist or adding them to the users spotify library.

This pr depends on #305 (as it is big enough, i hope 305 will make it in and make this pr a bit easier to review).

### CODE REVIEW

File: src/db.c [hunk 1]
```diff
@@ -2004,6 +2027,31 @@ db_files_get_count_bymatch(char *path)
 #undef Q_TMPL
 }
 
+int
+db_file_get_seekpos(uint32_t id)
```
Comment (ejurgensen) [inline src/db.c]: db_file_get_seekpos -> just use db_file_fetch_byid (or save the seek in the queue?)

File: src/db.c [hunk 2]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
 
+/* Queue */
 
-/* Inotify */
-int
-db_watch_clear(void)
-{
-  return db_query_run("DELETE FROM inotify;", 0, 0);
+static int
+queue_fetch_byitemid(struct db_queue_enum *queue_enum, uint32_t item_id, struct db_queue_item *queue_item, int keep_item);
```
Comment (ejurgensen) [inline src/db.c]: this forward doesn't seem required?

File: src/db.c [hunk 3]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
 }
 
+/*
+ * Returns the queue version from the admin table
+ *
+ * @return queue version
+ */
 int
-db_watch_add(struct watch_info *wi)
+db_queue_get_version()
```
Comment (ejurgensen) [inline src/db.c]: does db_queue_get_version need a function?

File: src/db.c [hunk 4]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
 {
-#define Q_TMPL "INSERT INTO inotify (wd, cookie, path) VALUES (%d, 0, '%q');"
-  char *query;
+  char *version;
+  int32_t plversion;
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wi->wd, wi->path);
+  plversion = 0;
+  version = db_admin_get("plversion");
```
Comment (ejurgensen) [inline src/db.c]: inconsistent naming? queue version or playlist version?

File: src/db.c [hunk 5]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+ *
+ * This function must be called after successfully modifying the queue table in order to send
+ * notification messages to the clients (e. g. dacp or mpd clients).
+ */
+static void
+queue_inc_version_and_notify()
 {
-#define Q_TMPL "DELETE FROM inotify WHERE wd = %d;"
-  char *query;
+  int plversion;
```
Comment (ejurgensen) [inline src/db.c]: naming

File: src/db.c [hunk 6]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+  char version[10];
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wd);
+  db_transaction_begin();
```
Comment (ejurgensen) [inline src/db.c]: why is a transaction required?

File: src/db.c [hunk 7]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
-db_watch_delete_bypath(char *path)
-{
-#define Q_TMPL "DELETE FROM inotify WHERE path = '%q';"
-  char *query;
+  plversion++;
+  ret = snprintf(version, sizeof(version), "%d", plversion);
+  if (ret >= sizeof(version))
+    {
+      DPRINTF(E_LOG, L_DB, "Error incrementing queue version. Could not convert version to string: %d\n", plversion);
+      db_transaction_rollback();
```
Comment (ejurgensen) [inline src/db.c]: isn't this rolling back a SELECT?

File: src/db.c [hunk 8]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+      return;
+    }
 
-  query = sqlite3_mprintf(Q_TMPL, path);
+  ret = db_admin_update("plversion", version);
+  if (ret < 0)
+    {
+      DPRINTF(E_LOG, L_DB, "Error incrementing queue version. Could not update version in admin table: %d\n", plversion);
+      db_transaction_rollback();
```
Comment (ejurgensen) [inline src/db.c]: rolling back a failed update?

File: src/db.c [hunk 9]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+  char version[10];
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wd);
+  db_transaction_begin();
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
-}
+  plversion = db_queue_get_version();
```
Comment (ejurgensen) [inline src/db.c]: don't know if a custom increment query might be an idea? plversion isn't used anyway (outside of this function)

File: src/db.c [hunk 10]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-int
-db_watch_delete_bycookie(uint32_t cookie)
-{
-#define Q_TMPL "DELETE FROM inotify WHERE cookie = %" PRIi64 ";"
+static int
+queue_add_mediafileinfo(struct db_media_file_info *dbmfi, int pos, int shuffle_pos)
```
Comment (ejurgensen) [inline src/db.c]: I think db_queue_add_file() would be more consistent naming

File: src/db.c [hunk 11]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+  return ret;
 
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-static int
-db_watch_get_byquery(struct watch_info *wi, char *query)
+/*
+ * Adds the files matching the given query to the queue after the item with the given item id
```
Comment (ejurgensen) [inline src/db.c]: good to have these, but I prefer them in the .h file

File: src/db.c [hunk 12]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
-  int ncols;
-  int i;
+  struct db_media_file_info dbmfi;
+  char *query;
+  int shuffle_pos;
+  int pos;
   int ret;
 
-  DPRINTF(E_DBG, L_DB, "Running query '%s'\n", query);
+  db_transaction_begin();
```
Comment (ejurgensen) [inline src/db.c]: why start the transaction before the updates? (also applies below)

File: src/db.c [hunk 13]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
-	      {
-		cookie = sqlite3_column_int64(stmt, i);
-		*ival = (cookie == INOTIFY_FAKE_COOKIE) ? 0 : cookie;
-	      }
-	    else
-	      *ival = sqlite3_column_int(stmt, i);
-	    break;
+      if (ret < 0)
+	{
+	  DPRINTF(E_DBG, L_DB, "Failed to add song  with id %s (%s) to queue\n", dbmfi.id, dbmfi.title);
```
Comment (ejurgensen) [inline src/db.c]: log level too low (also applies afterwards) + double space

File: src/db.c [hunk 14]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+db_queue_add_by_playlistid(int plid, char reshuffle, uint32_t item_id)
+{
+  struct query_params qp;
+  int ret;
+
+  memset(&qp, 0, sizeof(struct query_params));
+
+  qp.id = plid;
+  qp.type = Q_PLITEMS;
+  qp.offset = 0;
```
Comment (ejurgensen) [inline src/db.c]: maybe ok to omit some of this, even S_NONE and I_NONE, since you memset

File: src/db.c [hunk 15]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+  snprintf(buf, sizeof(buf), "f.id = %" PRIu32, id);
+  qp.filter = buf;
+
+  ret = db_queue_add_by_query(&qp, reshuffle, item_id);
+
+  return ret;
+}
+
+static int
+queue_enum_start(struct db_queue_enum *queue_enum)
```
Comment (ejurgensen) [inline src/db.c]: would a Q_QUEUE approach be possible?

File: src/db.c [hunk 16]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+      DPRINTF(E_DBG, L_DB, "End of queue enum results\n");
+      return 0;
+    }
+  else if (ret != SQLITE_ROW)
+    {
+      DPRINTF(E_LOG, L_DB, "Could not step: %s\n", sqlite3_errmsg(hdl));
+      return -1;
+    }
+
+  queue_item->item_id = (uint32_t)sqlite3_column_int(queue_enum->stmt, 0);
```
Comment (ejurgensen) [inline src/db.c]: yikes... didn't get any further than this (for now)

File: src/db.c [hunk 17]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+
+  return 0;
+}
+
+int
+db_queue_enum_start(struct db_queue_enum *queue_enum)
+{
+  int ret;
+
+  db_transaction_begin();
```
Comment (ejurgensen) [inline src/db.c]: ok I guess the transactions are for locking, but why now - there was no real locking before in db.c, and I think it had similar read + write sequences

File: src/db.c [hunk 18]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+
+  char *query;
+  int pos;
+
+  if (shuffle)
+    query = sqlite3_mprintf(Q_TMPL_SHUFFLE, file_id);
+  else
+    query = sqlite3_mprintf(Q_TMPL, file_id);
+
+  pos = db_get_count(query);
```
Comment (ejurgensen) [inline src/db.c]: guess db_get_count should be renamed to something else: db_get_one_int? db_get_one_result?

File: src/db.c [hunk 19]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+
+static int
+queue_fetch_byitemid(struct db_queue_enum *queue_enum, uint32_t item_id, struct db_queue_item *queue_item, int keep_item)
+{
+  int ret;
+
+  memset(queue_enum, 0, sizeof(struct db_queue_enum));
+  queue_enum->filter = sqlite3_mprintf("id = %d", item_id);
+
+  ret = queue_enum_start(queue_enum);
```
Comment (ejurgensen) [inline src/db.c]: I don't think it is nice to have a enum_start here, but require queue_enum_end of the caller

File: src/db.c [hunk 20]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+  queue_item = calloc(1, sizeof(struct db_queue_item));
+  if (!queue_item)
+    {
+      DPRINTF(E_LOG, L_DB, "Out of memory for queue_item\n");
+      return NULL;
+    }
+
+  db_transaction_begin();
+  ret = queue_fetch_byitemid(&queue_enum, item_id, queue_item, 1);
+  queue_enum_end(&queue_enum);
```
Comment (ejurgensen) [inline src/db.c]: might be ending a queue_enum that never started if ret < 0

File: src/db.c [hunk 21]
```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
...
+{
+  int ret;
+
+  memset(queue_enum, 0, sizeof(struct db_queue_enum));
+  if (shuffle)
+    queue_enum->filter = sqlite3_mprintf("shuffle_pos = %d", pos);
+  else
+    queue_enum->filter = sqlite3_mprintf("pos = %d", pos);
+
+  ret = queue_enum_start(queue_enum);
```
Comment (ejurgensen) [inline src/db.c]: as above, not a fan of splitting start and end

File: src/db.h [hunk 22]
```diff
@@ -373,6 +373,60 @@ struct directory_enum {
   sqlite3_stmt *stmt;
 };
 
+struct db_queue_item
+{
+  /* Item-Id is a unique id for this queue item. If the same item appears multiple
+     times in the queue each corresponding queue item has its own id. */
+  uint32_t item_id;
```
Comment (ejurgensen) [inline src/db.h]: just name it "id" (like it is in the db)

File: src/db.h [hunk 23]
```diff
@@ -373,6 +373,60 @@ struct directory_enum {
...
+  char *artist_sort;
+  char *album_sort;
+  char *album_artist_sort;
+
+  uint32_t year;
+  uint32_t track;
+  uint32_t disc;
+};
+
+struct db_queue_enum
```
Comment (ejurgensen) [inline src/db.h]: possible to use query_params instead?

File: src/db_init.h [hunk 24]
```diff
@@ -25,7 +25,7 @@
  * version of the database? If yes, then it is a minor upgrade, if no, then it
  * is a major upgrade. In other words minor version upgrades permit downgrading
  * forked-daapd after the database was upgraded. */
-#define SCHEMA_VERSION_MAJOR 19
+#define SCHEMA_VERSION_MAJOR 20
```
Comment (ejurgensen) [inline src/db_init.h]: isn't this actually a minor upgrade? adding a new table won't disturb old versions of forked-daapd

File: src/db.c (hunk 2 above)
Comment (chme) [inline src/db.c]: missed this during one of my rewrites, will remove it

File: src/db.c (hunk 3 above)
Comment (chme) [inline src/db.c]: With out this function the caller would have to convert the value to an int. I prefer to keep it. It encapsulates the knowledge how the queue-version is stored in the db.c file.

File: src/db.c (hunk 4 above)
Comment (chme) [inline src/db.c]: will change it to queue_version

File: src/db.c (hunk 6 above)
Comment (chme) [inline src/db.c]: The two sql statements (select and update) should be done in one transaction to avoid parallel updates. In this case parallel updates are not critical and the transaction is not strictly required. But without it, sqlite would open two implicit transactions (one for each statement), in my opinion it makes sense to avoid this overhead and simply open and close the transaction explicitly.

File: src/db.c (hunk 9 above)
Comment (chme) [inline src/db.c]: I would avoid custom sqlite extensions as long as they do not really add a benefit. If incrementing a value stored as an int is used some place else in forked-daapd, that may be the case. Otherwise i think the sql extension would make it harder to understand what happens.

File: src/db.c (hunk 8 above)
Comment (chme) [inline src/db.c]: Same as above, the transaction need to be closed either by end or rollback.

File: src/db.c (hunk 10 above)
Comment (chme) [inline src/db.c]: Would renaming it to queue_add_file be ok? (i think the "db_" prefix is used to mark public functions)

File: src/db.c (hunk 11 above)
Comment (chme) [inline src/db.c]: If you do not have a strong preference for putting the comments to .h file, i like to keep them at the implementation of a function in the .c file. I am using eclipse as editor and eclipse does not show the comments of the declaration (in the .h file) but instead shows the comments (+ function body) from the implementation (in the .c file).

File: src/db.c (hunk 12 above)
Comment (chme) [inline src/db.c]: It is necessary to wrap the reads and writes in a single transaction to guarantee that a parallel update does not change the values between reading and writing. This may lead to an inconsistent queue, with missing or duplicate entries for one position.

File: src/db.c (hunk 17 above)
Comment (chme) [inline src/db.c]: The queue is a bit different as the entries are not independent from each other. If the queue consists of three items the positions have to be 0, 1, 2. If concurrent threads try to update the queue, without transactions the result could be missing/duplicate entries for one position.

File: src/db.c (hunk 18 above)
Comment (chme) [inline src/db.c]: Should i do this in this pr or in a separate one (this will change otherwise unrelated code)?

File: src/db.c (hunk 19 above)
Comment (chme) [inline src/db.c]: Would it be enough to rename the function to something like "queue_start_and_fetch_byitemid"? It is only used internally in db.c.

My intention was to keep the number of memory allocations low. Therefor the statement must be explicitly closed by the caller (avoiding the strdups and explicitly freeing of the qeue_item).

File: src/db.h (hunk 22 above)
Comment (chme) [inline src/db.h]: ok

File: src/db_init.h (hunk 24 above)
Comment (chme) [inline src/db_init.h]: I have to say, that i am not a fan of the major/minor version for the db. In theory it would be possible to downgrade. But for that to work the user will have to change at least the db-version in the sqlite database. And it might not be possible to rerun the update on an already updated database. 

If you want, i can change it to a minor version update.

**Comment by chme** [comment]
Thanks for taking the time to review this pr. I will take a look at using the existing query params.

What is still missing in this pr, is the change for the raop metadata update in dmap_encode_file_metadata(...). This currently relies on the dmap-fields hashtable. I am not sure how to best adapt it to read from the queue instead of the files table. I avoided adding two structs for the queue table (like it is done e. g. for the files table with db_media_file_info and media_file_info). But the dmap-fields rely on getting all fields as a string. I am still working on it, trying to figure out what the best approach is.

If you have an idea/preference how to do it, that would be great.

**Comment by chme** [comment]
Commit fdd85cc builds raop metadata from the queue. It is untested (need to setup a raop client that requests metadata) and only sends a subset of the previously send metadata fields (i believe the important ones are still there), so maybe you can have a look if this is an acceptable approach. 

fyi: https://github.com/chme/forked-daapd/tree/mpdaddstream has additional changes that allow adding an internet stream directly to the queue without the need to put it in a playlist file.

File: src/db.c (hunk 11 above)
Comment (ejurgensen) [inline src/db.c]: Alright. I'm not an Eclipse user myself, but isn't this behavior odd? I mean most of the header files in /usr/include include documentation, so does Eclipse not show that? Like for the library headers, a caller shouldn't need to look in the implementation for anything.

File: src/db.c (hunk 18 above)
Comment (ejurgensen) [inline src/db.c]: I agree, a separate is probably better.

File: src/db.c (hunk 17 above)
Comment (ejurgensen) [inline src/db.c]: Ok, thanks for the explanation

File: src/db.c (hunk 9 above)
Comment (ejurgensen) [inline src/db.c]: Didn't mean an extension, just defining something like a Q_TMPL with an UPDATE and a version = version + 1.

**Comment by ejurgensen** [comment]
Thanks for the updates. I don't have any immediate ideas about the dmap thing, but I think your approach should be alright. I would think it unlikely that RAOP targets request more exotic metadata.

Good to hear you already have the new queueing features in action. I'm also looking forward to that global Spotify search & play :-)

File: src/db.c (hunk 11 above)
Comment (chme) [inline src/db.c]: Eclipse shows the docs for the implementation if available, if not it shows the docs for the declaration. I think, showing the most specific docs is common behavior for IDEs (if there were different implementations for a declaration with different docs, i expect the editor to show the docs of the implementation that is used during runtime). 

It is different for libraries, were it is necessary to have the docs in the header. But for an application there is no gain in putting them in the header files. 

(In Java there is an import docs statement, that adds the documentation from the declaration to the implementation, but i am not aware of a similar statement in c.)

File: src/db_init.h (hunk 24 above)
Comment (chme) [inline src/db_init.h]: The update is backwards compatible, i will change it to 19.01. 

Only the "INSERT INTO admin (key, value) VALUES ('queue_version', '0');" query leads to duplicate entries for the key "queue_version" if the update is run twice on the same db. The issue here is that the admin table does not define the key column as the primary key. An "insert or update" will also lead to duplicate entries because sqlite can not identify that the row with the same key value should be updated.

**Comment by chme** [comment]
I looked into using the generic query start function and i do not believe there is something to gain in using it for the queue table. It will just lead to an additional indirection. 

I would rather untangle the query start function and divide it into separate functions (one for each table that is queried). The caller knows which items it wants to read and using separate function will make this more explicit and easier to understand. (But i hope this is out of scope for this pr)

**Comment by chme** [comment]
edc609b addresses the asymmetry of the internal query start/fetch/end. I realized that the metadata fields are not used in the internal calls to e. g. queue_fetch_byitemid. That makes it possible to directly end the query (only int fields are accessed).

File: src/db.c (hunk 1 above)
Comment (chme) [inline src/db.c]: I do not have a strong opinion on this and will change it if you want, but it results in a (probably small) performance decrease, as fetching the whole media_file_info makes memory allocations for every string field in media_file_info.

I did consider adding it to the queue item, but i do not believe it is the right place (would require updating the queue by the player and there maybe places where it is not desirable to store it in the queue - e. g. if it is not the item where playback starts).

Comment (ejurgensen) [inline src/db.c]: I guess this is the trade-off between code reuse and optimization. I would go for reuse unless there is an actual performance bottleneck.

There are probably other places the entire mfi is fetched despite only a single int being looked for (like data_kind). I don't want db_file_get_x, _y and _z. Do you think it is possible to make a generic version, something like db_file_get_int(fieldref)? Then the fieldref probably shouldn't be the actual column name, but perhaps using the offset of mfi->field could work as an argument?

**Comment by ejurgensen** [comment]
Thanks for the PR, excellent work! Merged with master now. If there are further adjustments let me know.
